### PR TITLE
bugfix: correct node rank

### LINKED ISSUE #19436: Wrong node_rank caused index error
### Bug description

When you use multiple machines and multiple cards, incorrect calculation of node_RANK causes index out of range.

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

use streaming dataset in 8GPU * 2 Node

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

image

### PR DESCRIPTION
**Pull request by cauyxy**
## What does this PR do?

Corrected the logic of calculating rank from dividing by num_nodes to num_gpu_per_meachine

Fixes #19436 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ yes ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ yes ] Check that all items from **Before submitting** are resolved
- [ yes ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ yes ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19437.org.readthedocs.build/en/19437/

### CODE REVIEW

**Comment by cauyxy** [comment]
I think this is a common problem that almost everyone who uses multiple machines and multiple cards for training will encounter. Should a quick fix be released?

File: src/lightning/data/utilities/shuffle.py [hunk 1]
```diff
@@ -12,8 +12,9 @@ def _intra_node_chunk_shuffle(
     current_epoch: int,
 ) -> List[int]:
     chunk_indexes_per_nodes: Any = [[] for _ in range(distributed_env.num_nodes)]
+    process_per_node = distributed_env.world_size // distributed_env.num_nodes
```
Comment (cauyxy) [inline src/lightning/data/utilities/shuffle.py:15]: maybe add this attr to `_DistributedEnv ` ?

Comment (tchaton) [reply src/lightning/data/utilities/shuffle.py:15]: Yes, this is a good idea! Feel free to do so.

**Comment by tchaton** [comment]
Hey @cauyxy, 

Would you mind adding a test ? 

Best,
T.C

**Comment by cauyxy** [comment]
> Hey @cauyxy, great catch !
> 
> Would you mind adding a test ?
> 
> Best, T.C

Just like the previous one, it might take me some time if I were to do it myself. Can you help me out?😊

**Comment by cauyxy** [comment]
i have added test for dist training when gpu_per_node > node_num

**Review by tchaton — APPROVED** [review]
Thanks @cauyxy. This test fails on master and the fix is correct ! Thanks for your contribution !

**Pull request merged**